File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/misc/GuestModeController.java
package com.qimian233.ztool.hook.modules.systemui.misc;

import android.annotation.SuppressLint;
import android.content.Context;
import android.provider.Settings;

import com.qimian233.ztool.data.ScopeKeys;
import com.qimian233.ztool.hook.base.AppHookModule;

import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Method;

/**
 * 访客模式控制Hook模块
 * 修复系统UI中自动创建访客用户的逻辑
 * 当用户切换器被禁用时，阻止自动添加访客用户
 */
public class GuestModeController extends AppHookModule {

    public GuestModeController() {}

    @Override
    public String getModuleName() {
        return "guest_mode_controller";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{
                ScopeKeys.SYSTEM_UI.packageName
        };
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if (ScopeKeys.SYSTEM_UI.packageName.equals(packageName)) {
            hookGuestUserInteractor(classLoader);
        }
    }

    private void hookGuestUserInteractor(ClassLoader classLoader) {
        try {
            @SuppressLint("PrivateApi") Method isAllowedMethod = classLoader
                    .loadClass("com.android.systemui.user.domain.interactor.GuestUserInteractor")
                    .getDeclaredMethod("isDeviceAllowedToAddGuest");
            hookWithId(isAllowedMethod, "is_allowed", chain -> {
                // 获取应用上下文
                Context context = (Context) chain.getThisObject().getClass()
                        .getDeclaredField("applicationContext").get(chain.getThisObject());

                // 检查用户切换器是否启用
                int userSwitcherEnabled = Settings.Global.getInt(
                        context.getContentResolver(),
                        "user_switcher_enabled",
                        0
                );

                // 如果用户切换器被禁用，则不允许添加访客
                if (userSwitcherEnabled == 0) {
                    logger.debug("阻止自动添加访客用户 - 用户切换器已禁用");
                    return false;
                }
                return chain.proceed();
            });

            logger.info("成功Hook访客用户交互器");
        } catch (Throwable t) {
            logger.error("Hook访客用户交互器失败", t);
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/misc/GuestModeController.kt
package com.qimian233.ztool.hook.modules.systemui.misc

import android.annotation.SuppressLint
import android.content.Context
import android.provider.Settings
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam

/**
 * Guest mode control hook module.
 * Fixes the auto guest user creation logic in SystemUI.
 * Prevents automatically adding a guest user when the user switcher is disabled.
 */
class GuestModeController : AppHookModule() {
    override fun getModuleName(): String = PreferenceKeys.GUEST_MODE_CONTROLLER.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.SYSTEM_UI.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        hookGuestUserInteractor(classLoader)
    }

    private fun hookGuestUserInteractor(classLoader: ClassLoader) {
        try {
            @SuppressLint("PrivateApi") val isAllowedMethod = classLoader
                .loadClass("com.android.systemui.user.domain.interactor.GuestUserInteractor")
                .getDeclaredMethod("isDeviceAllowedToAddGuest")
            hookWithId(isAllowedMethod, "is_allowed") { chain ->
                // Get the application context
                val context = chain.thisObject.javaClass
                    .getDeclaredField("applicationContext").get(chain.thisObject) as Context?

                // Check whether the user switcher is enabled
                val userSwitcherEnabled = Settings.Global.getInt(
                    context!!.contentResolver,
                    "user_switcher_enabled",
                    0
                )

                // If the user switcher is disabled, do not allow adding a guest
                if (userSwitcherEnabled == 0) {
                    logger.debug("Blocked automatic guest user addition - user switcher disabled")
                    return@hookWithId false
                }
                chain.proceed()
            }

            logger.info("Successfully hooked GuestUserInteractor")
        } catch (t: Throwable) {
            logger.error("Failed to hook GuestUserInteractor", t)
        }
    }
}
